crates/storage/provider/src/providers/blockchain_provider.rs

```diff
@@ -175,9 +175,12 @@ impl<N: ProviderNodeTypes> BlockchainProvider2<N> {
         state: impl AsRef<BlockState>,
     ) -> ProviderResult<MemoryOverlayStateProvider> {
         let state = state.as_ref();
+        trace!(target: "providers::blockchain", "Calling anchor for fetching historical state provider");
         let anchor_hash = state.anchor().hash;
+        trace!(target: "providers::blockchain", ?anchor_hash, "Fetching historical provider for block hash");
         let latest_historical = self.database.history_by_block_hash(anchor_hash)?;
```

Comment:
can there still be a race condition here between the moment anchor hash was set on `BlockState` and the moment when we actually retrieve the state provider? does it matter here?

Reply:
>  can there still be a race condition here between the moment anchor hash was set on BlockState and the moment when we actually retrieve the state provider? does it matter here?

I think so, although only for reorgs afaict. When the DB state only moves forward this should not be a problem. Take for example a reorg which removes one block from the DB. The race is:
* we fetch the state
* anchor is removed from DB, replaced with another block
* now we call `history_by_block_hash` on something that does not exist

The only idea I can come up with to fix it, is to store `ExecutedBlock`s like we do `TrieUpdates` before finalization.